daffodil-core/src/main/scala/org/apache/daffodil/runtime1/TermRuntime1Mixin.scala

```diff
@@ -0,0 +1,189 @@
+package org.apache.daffodil.runtime1
+
+import org.apache.daffodil.xml.QNameBase
+import org.apache.daffodil.infoset.PartialNextElementResolver
+import org.apache.daffodil.infoset.SeveralPossibilitiesForNextElement
+import org.apache.daffodil.processors.ElementRuntimeData
+import org.apache.daffodil.infoset.NoNextElement
+import org.apache.daffodil.infoset.OnlyOnePossibilityForNextElement
+import org.apache.daffodil.api.WarnID
+import org.apache.daffodil.dsom.Term
+import org.apache.daffodil.dsom.ElementBase
+import org.apache.daffodil.dsom.PossibleNextElements
+import org.apache.daffodil.dsom.ModelGroup
+import org.apache.daffodil.dsom.DPathElementCompileInfo
+import org.apache.daffodil.dsom.PrefixLengthQuasiElementDecl
+import org.apache.daffodil.processors.TermRuntimeData
+import org.apache.daffodil.schema.annotation.props.gen.NilKind
+import org.apache.daffodil.schema.annotation.props.gen.Representation
+import org.apache.daffodil.schema.annotation.props.gen.LengthKind
+
+/**
+ * Part of Daffodil's Runtime 1 Streaming Unparser Support.
+ *
+ * When streaming in some representation of an infoset for unparsing,
+ * Daffodil (in Runtime 1) must resolve what element to construct.
+ * This is context specific, and given sharing of elements and groups,
+ * requires a runtime stack of TermRuntimeData, each stack entry has
+ * an associated PartialNextElementResolver.
+ *
+ * This object computes that PartialNextElementResolver based on the
+ * possible elements that can follow the current term.
+ */
+trait TermRuntime1Mixin { self: Term =>
+  import PossibleNextElements._
+
+  def termRuntimeData: TermRuntimeData
+
+  /**
+   * The PartialNextElementResolver is used to determine what infoset event comes next, and "resolves" which is to say
+   * determines the ElementRuntimeData for that infoset event. This can be used to construct the initial
+   * infoset from a stream of XML events.
+   */
+  final lazy val partialNextElementResolver: PartialNextElementResolver = {
+    val context = self
+    val possibles = possibleNextLexicalSiblingElementsInInfoset
+
+    //
+    // Annoying, but scala's immutable Map is not covariant in its first argument
+    // the way one would normally expect a collection to be.
+    // So Map[NamedQName, ElementRuntimeData] is not a subtype of Map[QNameBase, ElementRuntimeData]
+    // So we need a cast upward to Map[QNameBase,ElementRuntimeData]
+    //
+    val schemaFileLocation = context.schemaFileLocation
+    val (sibs, isRequiredInInfoset) = possibles match {
+      case Closed(sibs) => (sibs, true)
+      case Open(sibs) => (sibs, false)
+    }
+    val eltMap = sibs.map {
+      sib => (sib.e.namedQName, sib.e.erd)
+    }.toMap.asInstanceOf[Map[QNameBase, ElementRuntimeData]]
+    val resolver = eltMap.size match {
+      case 0 => new NoNextElement(schemaFileLocation)
```

Comment:
Do we actually use NoNextElement resolver, or is it redundant with Closed state of possibles (which turns into the flag isRequiredInInfoset being true, and at runtime the nextElement method of NextElementResolver deals with Closed nature.

Reply:
It is not redundant. NoNextElement resolver is used for a complex element with no element children. It is also used for a sequence that is the model group of a complex type.